### PR TITLE
COL-365, deploy script now supports deploy-by-tag

### PR DESCRIPTION
**Pull request by johncrossman**
https://jira.ets.berkeley.edu/jira/browse/COL-365
https://jira.ets.berkeley.edu/jira/browse/COL-486

The script will:
* cd to base dir of local repo
* log important stuff to time-stamped log file
* validate, report errors (e.g., no such tag)
* give user more info via console
* pause before blowing away local changes

### CODE REVIEW

File: deploy/deploy.sh [hunk 1]
```diff
@@ -1,57 +1,100 @@
...
-# Get the remote and branch that should be
-# deployed from the provided environment variables
-TARGET_REMOTE=${REMOTE:-origin}
-TARGET_BRANCH=${BRANCH:-master}
+# Go to base directory of local git repo
+cd $(dirname "${BASH_SOURCE[0]}")/..
+base_directory="${PWD}"
+
+echo_usage() {
+  echo; echo "USAGE"; echo "  ${0} [-t tag] [-b branch]"; echo
```
Comment (paulkerschen) [inline deploy/deploy.sh]: @johncrossman Should -r for remote be documented here as well?

File: deploy/deploy.sh [hunk 2]
```diff
@@ -1,57 +1,100 @@
...
+  esac
+done
+
+# Validation
+[[ "${git_tag}" || "${git_branch}" ]] || { echo; echo "[ERROR] You must specify branch or tag."; echo_usage; exit 1; }
+[[ "${git_tag}" && "${git_branch}" ]] && { echo; echo "[ERROR] Specify branch or tag but NOT both."; echo_usage; exit 1; }
 
-# Clear any local changes present in the branch
+echo; echo "WARNING! In two seconds we will clear local changes with git reset. Control-c to abort."; echo
+sleep 2.5
```
Comment (paulkerschen) [inline deploy/deploy.sh]: I like that the user gets an extra half second to avert disaster!

File: deploy/deploy.sh [hunk 3]
```diff
@@ -1,57 +1,100 @@
...
+  git checkout tags/${git_tag} || { echo "[ERROR] Unknown Git tag: ${git_tag}"; echo; exit 1 }
+fi
+
+echo "The Git checkout is complete. Next, install." | ${logger}
 
 # Check out the branch or tag. If a tag is being deployed, the git HEAD will point to a commit and
 # will end up in a "detached" state. As we shouldn't be committing on deployed code, this is considered OK
-git checkout $TARGET_REMOTE/$TARGET_BRANCH
-git branch -D tmp
+# :TODO: Is a tmp branch necessary?
```
Comment (paulkerschen) [inline deploy/deploy.sh]: @johncrossman I've wondered the same thing. Worth a minor-priority JIRA to investigate?

File: deploy/deploy.sh (hunk 1 above)
Comment (johncrossman) [inline deploy/deploy.sh]: Nice catch, @paulkerschen  Amending.

**Comment by paulkerschen** [comment]
Nice looking script, @johncrossman. Looks good to go on amendment and re-Travis.

File: deploy/deploy.sh [hunk 4]
```diff
@@ -1,57 +1,107 @@
...
+  echo; echo "USAGE"; echo "  ${0} [-r remote] [-b branch] [-t tag]"; echo
+  echo "Deploy the latest SuiteC code from a specified branch or tag."; echo
+  echo "Set environment variable DOCUMENT_ROOT to the Apache directory which will serve SuiteC static files."; echo
+  echo "Common usages:"; echo
+  echo "   # Deploy Arthur's branch"
+  echo "   ${0} -r arthur_guinness -b irish_dry_stout"; echo
+  echo "   # Deploy qa branch using default remote (origin)"
+  echo "   ${0} -b qa"; echo
+  echo "   # Deploy tag 1.6"
+  echo "   ${0} -t 1.6"; echo
```
Comment (johncrossman) [inline deploy/deploy.sh]: @paulkerschen Usage info is fixed. The examples give a sense of what is optional and what is required.

**Comment by johncrossman** [comment]
@paulkerschen Hold off on merge.  I have one fix to apply.

**Comment by johncrossman** [comment]
@paulkerschen amendments are in. The local branch is given a more meaningful name. You'll also notice `log()` function for more concise log statements.

**Comment by paulkerschen** [comment]
Très bien @johncrossman.

**Pull request merged**